File: src/org/epstudios/epmobile/Qtc.java
/*  EP Mobile -- Mobile tools for electrophysiologists
    Copyright (C) 2011 EP Studios, Inc.
    www.epstudiossoftware.com

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.epstudios.epmobile;

import org.epstudios.epmobile.QtcCalculator.QtcFormula;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

public class Qtc extends EpActivity implements OnClickListener {
	private enum IntervalRate {
		INTERVAL, RATE
	};

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.qtc);

		View calculateQtcButton = findViewById(R.id.calculate_qtc_button);
		calculateQtcButton.setOnClickListener(this);
		View clearButton = findViewById(R.id.clear_button);
		clearButton.setOnClickListener(this);

		intervalRateSpinner = (Spinner) findViewById(R.id.interval_rate_spinner);
		qtcTextView = (TextView) findViewById(R.id.calculated_qtc);
		rrEditText = (EditText) findViewById(R.id.rrEditText);
		qtEditText = (EditText) findViewById(R.id.qtEditText);
		qtcFormulaTextView = (TextView) findViewById(R.id.qtc_formula);

		getPrefs();
		setAdapters();

		clearEntries();

	}

	private Spinner intervalRateSpinner;
	private TextView qtcTextView;
	private EditText rrEditText;
	private EditText qtEditText;
	private TextView qtcFormulaTextView;
	private String qtcFormula;
	private OnItemSelectedListener itemListener;

	private int qtcUpperLimit;
	private final static int QTC_UPPER_LIMIT = 440;
	private final static int INTERVAL_SELECTION = 0;
	private final static int RATE_SELECTION = 1;

	private IntervalRate defaultIntervalRateSelection = IntervalRate.INTERVAL;

	public void onClick(View v) {
		switch (v.getId()) {
		case R.id.calculate_qtc_button:
			calculateQtc();
			break;
		case R.id.clear_button:
			clearEntries();
			break;
		}
	}

	private void setAdapters() {
		ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(
				this, R.array.interval_rate_labels,
				android.R.layout.simple_spinner_item);
		adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
		intervalRateSpinner.setAdapter(adapter);
		if (defaultIntervalRateSelection.equals(IntervalRate.INTERVAL))
			intervalRateSpinner.setSelection(INTERVAL_SELECTION);
		else
			intervalRateSpinner.setSelection(RATE_SELECTION);
		itemListener = new OnItemSelectedListener() {
			public void onItemSelected(AdapterView<?> parent, View v,
					int position, long id) {
				updateIntervalRateSelection();
			}

			public void onNothingSelected(AdapterView<?> parent) {
				// do nothing
			}

		};

		intervalRateSpinner.setOnItemSelectedListener(itemListener);

	}

	private void updateIntervalRateSelection() {
		IntervalRate intervalRateSelection = getIntervalRateSelection();
		if (intervalRateSelection.equals(IntervalRate.INTERVAL))
			rrEditText.setHint(getString(R.string.rr_hint));
		else
			rrEditText.setHint(getString(R.string.hr_hint));
	}

	private IntervalRate getIntervalRateSelection() {
		String result = intervalRateSpinner.getSelectedItem().toString();
		if (result.startsWith("RR"))
			return IntervalRate.INTERVAL;
		else
			return IntervalRate.RATE;

	}

	private void showQtcFormula() {
		qtcFormulaTextView.setText("QTc formula used was " + qtcFormula);
	}

	private void calculateQtc() {
		CharSequence rrText = rrEditText.getText();
		CharSequence qtText = qtEditText.getText();
		IntervalRate intervalRateSelection = getIntervalRateSelection();
		try {
			int rr = Integer.parseInt(rrText.toString());
			if (intervalRateSelection.equals(IntervalRate.RATE))
				rr = 60000 / rr;
			int qt = Integer.parseInt(qtText.toString());
			getPrefs();
			showQtcFormula();
			QtcFormula formula;
			if (qtcFormula.equals("BAZETT"))
				formula = QtcFormula.BAZETT;
			else if (qtcFormula.equals("FRIDERICIA"))
				formula = QtcFormula.FRIDERICIA;
			else if (qtcFormula.equals("SAGIE"))
				formula = QtcFormula.SAGIE;
			else if (qtcFormula.equals("HODGES"))
				formula = QtcFormula.HODGES;
			else
				formula = QtcFormula.BAZETT;
			Toast.makeText(this, "QTc Formula is " + qtcFormula, 3000).show();
			int qtc = QtcCalculator.calculate(rr, qt, formula);
			qtcTextView.setText("QTc = " + String.valueOf(qtc) + " msec");
			if (qtc >= qtcUpperLimit)
				qtcTextView.setTextColor(Color.RED);
			else
				qtcTextView.setTextColor(Color.GREEN);
		} catch (NumberFormatException e) {
			qtcTextView.setText("Invalid!");
			qtcTextView.setTextColor(Color.RED);
		}
	}

	private void clearEntries() {
		rrEditText.setText(null);
		qtEditText.setText(null);
		qtcTextView.setText(getString(R.string.qtc_result_label));
		qtcTextView.setTextColor(Color.LTGRAY);
		qtcFormulaTextView.setText(null);
		rrEditText.requestFocus();
	}

	private void getPrefs() {
		SharedPreferences prefs = PreferenceManager
				.getDefaultSharedPreferences(getBaseContext());
		qtcFormula = prefs.getString("qtc_formula", "BAZETT");
		String intervalRatePreference = prefs.getString("interval_rate",
				"INTERVAL");
		if (intervalRatePreference.equals("INTERVAL"))
			defaultIntervalRateSelection = IntervalRate.INTERVAL;
		else
			defaultIntervalRateSelection = IntervalRate.RATE;
		String s = prefs.getString("maximum_qtc", "");
		try {
			qtcUpperLimit = Integer.parseInt(s);
		} catch (NumberFormatException e) {
			qtcUpperLimit = QTC_UPPER_LIMIT;
			SharedPreferences.Editor editor = prefs.edit();
			editor.putString("maximum_qtc", String.valueOf(QTC_UPPER_LIMIT));
			editor.commit();
		}
	}

}

File: app/src/main/java/org/epstudios/epmobile/features/calculators/ui/Qtc.kt
package org.epstudios.epmobile.features.calculators.ui

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.preference.PreferenceManager
import org.epstudios.epmobile.R
import org.epstudios.epmobile.core.data.QtcCalculator
import org.epstudios.epmobile.core.data.QtcCalculator.QtcFormula
import org.epstudios.epmobile.core.data.Reference
import org.epstudios.epmobile.core.ui.base.EpActivity
import org.epstudios.epmobile.databinding.QtcBinding

class Qtc : EpActivity() {
    private enum class IntervalRate { INTERVAL, RATE }

    private lateinit var binding: QtcBinding

    private var qtcUpperLimit: Int = QTC_UPPER_LIMIT
    private var defaultIntervalRateSelection = IntervalRate.INTERVAL
    private var qtcFormula: String = BAZETT

    companion object {
        private const val QTC_UPPER_LIMIT = 440
        private const val BAZETT = "BAZETT"
        private const val FRIDERICIA = "FRIDERICIA"
        private const val SAGIE = "SAGIE"
        private const val HODGES = "HODGES"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = QtcBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupInsets(binding.qtcRootView)
        initToolbar()

        binding.calculateButtonsLayout.calculateButton.setOnClickListener { calculateQtc() }
        binding.calculateButtonsLayout.clearButton.setOnClickListener { clearEntries() }

        getPrefs()
        setAdapters()
        setFormulaAdapters()

        clearEntries()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            val parentActivityIntent = Intent(this, CalculatorList::class.java)
            parentActivityIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            startActivity(parentActivityIntent)
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setAdapters() {
        val adapter = ArrayAdapter.createFromResource(
            this, R.array.interval_rate_labels,
            R.layout.dropdown_menu_item
        )
        binding.intervalRateSpinner.setAdapter(adapter)
        val initialPosition = if (defaultIntervalRateSelection == IntervalRate.INTERVAL) 0 else 1
        binding.intervalRateSpinner.setText(adapter.getItem(initialPosition), false)
        updateIntervalRateSelection()

        binding.intervalRateSpinner.setOnItemClickListener { _, _, _, _ ->
            updateIntervalRateSelection()
        }
    }

    private fun setFormulaAdapters() {
        val adapter = ArrayAdapter.createFromResource(
            this, R.array.formula_names,
            R.layout.dropdown_menu_item
        )
        binding.qtcFormulaSpinner.setAdapter(adapter)
        val formula = when (getQtcFormula(qtcFormula)) {
            QtcFormula.BAZETT -> 0
            QtcFormula.FRIDERICIA -> 1
            QtcFormula.SAGIE -> 2
            QtcFormula.HODGES -> 3
        }
        binding.qtcFormulaSpinner.setText(adapter.getItem(formula), false)

        binding.qtcFormulaSpinner.setOnItemClickListener { _, _, position, _ ->
            qtcFormula = when (position) {
                0 -> BAZETT
                1 -> FRIDERICIA
                2 -> SAGIE
                else -> HODGES
            }
        }
    }

    private fun updateIntervalRateSelection() {
        val hint = if (getIntervalRateSelection() == IntervalRate.INTERVAL) {
            getString(R.string.rr_hint)
        } else {
            getString(R.string.hr_hint)
        }
        binding.rrInputLayout.hint = hint
    }

    private fun getIntervalRateSelection(): IntervalRate {
        return if (binding.intervalRateSpinner.text.toString().startsWith("RR")) {
            IntervalRate.INTERVAL
        } else {
            IntervalRate.RATE
        }
    }

    private fun showQtcFormula() {
        binding.qtcFormula.text = getString(R.string.qtc_formula_used, qtcFormula)
    }

    private fun calculateQtc() {
        val rrText = binding.rrEditText.text.toString()
        val qtText = binding.qtEditText.text.toString()

        try {
            var rr = rrText.toInt()
            if (getIntervalRateSelection() == IntervalRate.RATE) {
                rr = (60000.0 / rr).toInt()
            }
            val qt = qtText.toInt()

            showQtcFormula()
            val formula = getQtcFormula(qtcFormula)
            Toast.makeText(this, "QTc Formula is $qtcFormula", Toast.LENGTH_LONG).show()
            val qtc = QtcCalculator.calculate(rr, qt, formula)
            binding.calculatedQtc.text = getString(R.string.qtc_result, qtc.toString())

            if (qtc >= qtcUpperLimit) {
                binding.calculatedQtc.setTextAppearance(R.style.TextAppearance_Calculator_Error)
            } else {
                binding.calculatedQtc.setTextAppearance(R.style.TextAppearance_Calculator_Result)
            }
        } catch (e: NumberFormatException) {
            binding.calculatedQtc.text = getString(R.string.invalid_warning)
            binding.calculatedQtc.setTextAppearance(R.style.TextAppearance_Calculator_Error)
        }
    }

    private fun getQtcFormula(name: String): QtcFormula {
        return when (name) {
            FRIDERICIA -> QtcFormula.FRIDERICIA
            SAGIE -> QtcFormula.SAGIE
            HODGES -> QtcFormula.HODGES
            else -> QtcFormula.BAZETT
        }
    }

    private fun clearEntries() {
        binding.rrEditText.text = null
        binding.qtEditText.text = null
        binding.calculatedQtc.text = getString(R.string.qtc_result_label)
        binding.calculatedQtc.setTextAppearance(R.style.TextAppearance_Calculator_Result)
        binding.qtcFormula.text = null
        binding.rrEditText.requestFocus()
    }

    private fun getPrefs() {
        val prefs = PreferenceManager.getDefaultSharedPreferences(baseContext)
        qtcFormula = prefs.getString("qtc_formula", BAZETT) ?: BAZETT
        val intervalRatePreference = prefs.getString("interval_rate", "INTERVAL")
        defaultIntervalRateSelection = if (intervalRatePreference == "INTERVAL") IntervalRate.INTERVAL else IntervalRate.RATE
        val s = prefs.getString("maximum_qtc", "")
        try {
            qtcUpperLimit = s?.toInt() ?: QTC_UPPER_LIMIT
        } catch (e: NumberFormatException) {
            qtcUpperLimit = QTC_UPPER_LIMIT
            val editor = prefs.edit()
            editor.putString("maximum_qtc", QTC_UPPER_LIMIT.toString())
            editor.apply()
        }
    }

    override fun hideReferenceMenuItem() = false

    override fun showActivityReference() {
        val referenceBazett = Reference(this, R.string.qtc_calculator_reference_bazett, R.string.qtc_calculator_link_bazett)
        val referenceFridericia = Reference(this, R.string.qtc_calculator_reference_fridericia, R.string.qtc_calculator_link_fridericia)
        val referenceSagie = Reference(this, R.string.qtc_calculator_reference_sagie, R.string.qtc_calculator_link_sagie)
        val referenceHodges = Reference(getString(R.string.qtc_calculator_reference_hodges), null)
        val referenceQtcLimits = Reference(this, R.string.qtc_limits_reference, R.string.qtc_limits_link)
        val references = arrayOf(referenceBazett, referenceFridericia, referenceSagie, referenceHodges, referenceQtcLimits)
        showReferenceAlertDialog(references)
    }
}
